### PR TITLE
PUBDEV-6754: provide actual ntree value

### PR DESCRIPTION
**Pull request by koniecsveta**

### CODE REVIEW

File: h2o-py/h2o/estimators/gbm.py [hunk 1]
```diff
@@ -953,3 +953,10 @@ def check_constant_response(self, check_constant_response):
         self._parms["check_constant_response"] = check_constant_response
 
 
+    """
+      Returns actual number of trees in GBM model (whether CV is enabled or not).
```
Comment (michalkurka) [inline h2o-py/h2o/estimators/gbm.py]: How is this related to CV?

I think we can mention early stopping instead.

File: h2o-r/h2o-package/R/gbm.R [hunk 2]
```diff
@@ -317,5 +317,8 @@ h2o.gbm <- function(x,
 
   # Error check and build model
   model <- .h2o.modelJob('gbm', parms, h2oRestApiVersion=3, verbose=verbose)
+
+  # Provide actual ntree value 
+  model@model$ntree_actual <- model@model$model_summary['number_of_trees'][,1]
```
Comment (michalkurka) [inline h2o-r/h2o-package/R/gbm.R]: Should we not provide a similar convenience method as we have in Python? Especially for the parity of APIs.

File: h2o-py/tests/testdir_algos/gbm/pyunit_PUBDEV-6754_gbm_ntree_actual.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  prostate_gbm = H2OGradientBoostingEstimator(nfolds=5,ntrees=ntrees_original, distribution="bernoulli", stopping_metric="MSE", stopping_tolerance=0.01, stopping_rounds=5)
+  prostate_gbm.train(x=list(range(2,9)), y=1, training_frame=prostate)
+  
+  print("\n")
+  print("number of trees set by user before building the model is:")
+  print(ntrees_original)
+  print("number of trees built with early-stopping is:")
+  print(prostate_gbm.ntree_actual())
+  
+  assert prostate_gbm.ntree_actual() == prostate_gbm._model_json['output']['model_summary']['number_of_trees'][0]
```
Comment (michalkurka) [inline h2o-py/tests/testdir_algos/gbm/pyunit_PUBDEV-6754_gbm_ntree_actual.py]: I would also suggest checking that the model actually stopped early. That is actual number of trees < 1000.

File: h2o-py/h2o/estimators/gbm.py (hunk 1 above)
Comment (koniecsveta) [reply h2o-py/h2o/estimators/gbm.py]: It is taken from the ticket description. Will fix.

File: h2o-bindings/bin/custom/python/gen_gbm.py [hunk 4]
```diff
@@ -3,6 +3,19 @@ def update_param(name, param):
         param['values'].remove('ordinal')
         return param
     return None  # param untouched
+ 
+def class_extensions():
```
Comment (honzasterba) [inline h2o-bindings/bin/custom/python/gen_gbm.py]: do we want the same for other tree algos as well? drf, xgboost, isofor?

File: h2o-bindings/bin/custom/python/gen_gbm.py [hunk 5]
```diff
@@ -3,6 +3,19 @@ def update_param(name, param):
...
+    """
+      Returns actual number of trees in GBM model (whether CV is enabled or not).
+
+      Type: ``float``
+      """
+    def ntree_actual(self):
```
Comment (honzasterba) [inline h2o-bindings/bin/custom/python/gen_gbm.py]: I dont like the name. Everywhere we use ntrees, my suggestions:
- ntrees_actual
- ntrees_built

Comment (koniecsveta) [reply h2o-bindings/bin/custom/python/gen_gbm.py]: Taken from the ticket description. Will change :)

File: h2o-bindings/bin/custom/python/gen_gbm.py (hunk 4 above)
Comment (koniecsveta) [reply h2o-bindings/bin/custom/python/gen_gbm.py]: not specified on the ticket but clarifying in #h2o-3-r-client

File: h2o-r/h2o-package/R/gbm.R (hunk 2 above)
Comment (koniecsveta) [reply h2o-r/h2o-package/R/gbm.R]: After clarification with @ledell and @Pscheidl this version is considered to be sufficient since we don’t have accessor method for all the other model parameters (just a few) + it will be implemented for all tree-based algorithms.

Comment (michalkurka) [reply h2o-r/h2o-package/R/gbm.R]: Thanks @koniecsveta

File: h2o-py/tests/testdir_algos/sharedtree/pyunit_PUBDEV-6754_ntrees_actual_for_tree_algos.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+from h2o.estimators.gbm import H2OGradientBoostingEstimator
+from h2o.estimators.isolation_forest import H2OIsolationForestEstimator
+from h2o.estimators.random_forest import H2ORandomForestEstimator
+from h2o.estimators.xgboost import H2OXGBoostEstimator
+
+def tree_algos_ntree_actual():
+  prostate = h2o.import_file(path=pyunit_utils.locate("smalldata/logreg/prostate.csv"))
+  prostate[1] = prostate[1].asfactor()
+  prostate.summary()
+  ntrees_original = 1000
```
Comment (honzasterba) [inline h2o-py/tests/testdir_algos/sharedtree/pyunit_PUBDEV-6754_ntrees_actual_for_tree_algos.py]: ideally I would chose a much lower number here, so that the test runs as fast as possible (same for R test)

File: h2o-py/h2o/estimators/gbm.py (hunk 1 above)
Comment (koniecsveta) [reply h2o-py/h2o/estimators/gbm.py]: fixed

File: h2o-py/tests/testdir_algos/gbm/pyunit_PUBDEV-6754_gbm_ntree_actual.py (hunk 3 above)
Comment (koniecsveta) [reply h2o-py/tests/testdir_algos/gbm/pyunit_PUBDEV-6754_gbm_ntree_actual.py]: done

File: h2o-bindings/bin/custom/python/gen_gbm.py (hunk 4 above)
Comment (koniecsveta) [reply h2o-bindings/bin/custom/python/gen_gbm.py]: added for all tree algos

File: h2o-bindings/bin/custom/python/gen_gbm.py (hunk 5 above)
Comment (koniecsveta) [reply h2o-bindings/bin/custom/python/gen_gbm.py]: done

File: h2o-py/tests/testdir_algos/sharedtree/pyunit_PUBDEV-6754_ntrees_actual_for_tree_algos.py (hunk 6 above)
Comment (koniecsveta) [reply h2o-py/tests/testdir_algos/sharedtree/pyunit_PUBDEV-6754_ntrees_actual_for_tree_algos.py]: The test runs faster because of early stopping. trees_original is 1000 to give enough space for build to be early stopped and to be illustrative / that there is a difference between trees_original and trees_actual (so it builds ~20 trees for each model).

File: h2o-py/h2o/model/model_base.py [hunk 7]
```diff
@@ -390,6 +390,20 @@ def scoring_history(self):
         print("No score history for this model")
 
 
+    def ntrees_actual(self):
+        """
+        Returns actual number of trees in GBM model. If early stopping enabled, GBM can reset the ntrees value.
```
Comment (honzasterba) [inline h2o-py/h2o/model/model_base.py]: mentions GBM when in facts applies to other algos

File: h2o-py/h2o/model/model_base.py [hunk 8]
```diff
@@ -390,6 +390,20 @@ def scoring_history(self):
         print("No score history for this model")
 
 
+    def ntrees_actual(self):
```
Comment (honzasterba) [inline h2o-py/h2o/model/model_base.py]: I am not sure I like this approach, this adds this method to models where it does not make sense. @michalkurka what do you think?

Comment (michalkurka) [reply h2o-py/h2o/model/model_base.py]: Good point @honza, this is however a common approach - this class already does have algo-specific method.

I would keep this for now.

Comment (michalkurka) [reply h2o-py/h2o/model/model_base.py]: If you want to make a jira - I think we can fix this in the future (we can eg. dynamically remove methods from the instances).

Comment (honza) [reply h2o-py/h2o/model/model_base.py]: :wave:

File: h2o-py/h2o/model/model_base.py (hunk 7 above)
Comment (koniecsveta) [reply h2o-py/h2o/model/model_base.py]: fixed
